File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/image/BitmapImage.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.image;

import android.graphics.Bitmap;
import android.media.Image;
import androidx.annotation.Nullable;

public class BitmapImage extends AbstractImage {

  private final Bitmap mBitmap;

  public BitmapImage(Bitmap bitmap) {
    mBitmap = bitmap;
  }

  @Override
  public float getNaturalWidth() {
    return mBitmap.getWidth();
  }

  @Override
  public float getNaturalHeight() {
    return mBitmap.getHeight();
  }

  @Override
  public Bitmap getBitmap() {
    return mBitmap;
  }

  @Nullable
  @Override
  public Image getImage() {
    return null;
  }

  @Override
  public int getImageRotationDegrees() {
    return 0;
  }

  @Override
  public void close() throws Exception {
    mBitmap.recycle();
  }
}

File: react-native-pytorch-core/ios/Image/BitmapImage.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public class BitmapImage: IImage {
    private var mBitmap: CGImage?
    private var ciImage: CIImage?
    public var width: CGFloat
    public var height: CGFloat

    init(image: CGImage) {
        self.mBitmap = image
        self.width = CGFloat(truncating: image.width as NSNumber)
        self.height = CGFloat(truncating: image.height as NSNumber)
    }

    init(image: CIImage) {
        self.ciImage = image
        self.width = image.extent.width
        self.height = image.extent.height
    }

    public func getPixelDensity() -> CGFloat {
        return 1.0
    }

    public func getWidth() -> CGFloat {
        return self.width
    }

    public func getHeight() -> CGFloat {
        return self.height
    }

    public func getNaturalWidth() -> Int {
        return Int(self.width)
    }

    public func getNaturalHeight() -> Int {
        return Int(self.height)
    }

    public func scale(sx: CGFloat, sy: CGFloat) throws -> IImage {
        return try ImageUtils.scale(image: self, sx: sx, sy: sy)
    }

    public func getBitmap() -> CGImage? {
        if let bitmap = mBitmap {
            return bitmap
        } else {
            let context = CIContext()
            if let ciImage = ciImage, let bitmap = context.createCGImage(ciImage, from: ciImage.extent) {
                mBitmap = bitmap
                return bitmap
            } else {
                return nil
            }
        }
    }

    public func close() throws {
        // TODO(T94684939)
    }
}
